File: backend/src/main/java/es/udc/paproject/backend/model/services/TrainingService.java
package es.udc.paproject.backend.model.services;

import java.time.LocalDateTime;
import java.util.List;

import es.udc.paproject.backend.model.entities.Training;
import es.udc.paproject.backend.model.exceptions.InstanceNotFoundException;
import es.udc.paproject.backend.model.exceptions.StartDateAfterEndDateException;
import es.udc.paproject.backend.model.exceptions.UsedTrainingException;

public interface TrainingService {
    
    //Busco un SeasonTeam con el teamId y cojo ese
    Training addTraining(Long teamId, Long seasonId, LocalDateTime trainingDate, String durationMinutes, String description, String objective) throws InstanceNotFoundException;

    void addPlayerToTraining(Long trainingId, Long playerId) throws InstanceNotFoundException;

    Training findTrainingById(Long trainingId) throws InstanceNotFoundException;

    List<Training> findTrainingsByUserId(Long userId) throws InstanceNotFoundException;

    List<Training> findTrainingsByPlayerId(Long playerId) throws InstanceNotFoundException;

    List<Training> findTrainingsByTwoDatesAndTeamIdOrSeasonId(Long userId, Long teamId, Long seasonId, LocalDateTime startDate, LocalDateTime endDate) throws StartDateAfterEndDateException, InstanceNotFoundException;

    List<Training> findTrainingsByTeamId(Long teamId) throws InstanceNotFoundException;

    List<Training> findTrainingsBySeasonId(Long seasonId) throws InstanceNotFoundException;
    //no pasar al front. Usar las dos anteriores
    List<Training> findTrainingsByTeamIdAndSeasonId(Long teamId, Long seasonId) throws InstanceNotFoundException;

    void removeTraining(Long trainingId) throws InstanceNotFoundException, UsedTrainingException;
    
    void removePlayerToTraining(Long playerId, Long trainingId) throws InstanceNotFoundException, UsedTrainingException;

    Training updateTraining(Long trainingId, LocalDateTime trainingDate, String durationMinutes, String description, String objective) throws InstanceNotFoundException;
}

File: frontend/src/backend/trainingService.js
import {config, appFetch} from './appFetch';

export const findTrainingById = (trainingId, onSuccess, onErrors) =>
    appFetch(`/trainings/${trainingId}`, config('GET'), onSuccess, onErrors);

export const findTrainingsByUserId = (onSuccess, onErrors) =>
    appFetch('/trainings/user', config('GET'), onSuccess, onErrors);

export const findTrainingsByPlayerId = (playerId, onSuccess, onErrors) =>
    appFetch(`/trainings/${playerId}/player`, config('GET'), onSuccess, onErrors);

export const findTrainingsByTwoDatesAndTeamIdOrSeasonId = (teamId, seasonId, startDate, endDate, onSuccess, onErrors) =>
    appFetch(`/trainings/dates?teamId=${teamId}&seasonId=${seasonId}&startDate=${startDate}&endDate=${endDate}`, config('GET'), onSuccess, onErrors);

export const findTrainingsByTwoDates = (startDate, endDate, onSuccess, onErrors) =>
    appFetch(`/trainings/dates?startDate=${startDate}&endDate=${endDate}`, config('GET'), onSuccess, onErrors);

export const findTrainingsByTwoDatesAndSeasonId = (seasonId, startDate, endDate, onSuccess, onErrors) =>
    appFetch(`/trainings/dates?seasonId=${seasonId}&startDate=${startDate}&endDate=${endDate}`, config('GET'), onSuccess, onErrors);

export const findTrainingsByTwoDatesAndTeamId = (teamId, startDate, endDate, onSuccess, onErrors) =>
    appFetch(`/trainings/dates?teamId=${teamId}&startDate=${startDate}&endDate=${endDate}`, config('GET'), onSuccess, onErrors);

export const findTrainingsByTeamId = (teamId, onSuccess, onErrors) =>
    appFetch(`/trainings/${teamId}/team`, config('GET'), onSuccess, onErrors);

export const findTrainingsBySeasonId = (seasonId, onSuccess, onErrors) =>
    appFetch(`/trainings/${seasonId}/season`, config('GET'), onSuccess, onErrors);

export const addTraining = (teamId, seasonId, trainingDate, durationMinutes, description, objective, onSuccess, onErrors) =>
    appFetch(`/trainings?teamId=${teamId}&seasonId=${seasonId}&trainingDate=${trainingDate}&durationMinutes=${durationMinutes}&description=${description}&objective=${objective}`, config('POST'), onSuccess, onErrors);

export const addTrainingWithTeam = (teamId, trainingDate, durationMinutes, description, objective, onSuccess, onErrors) =>
    appFetch(`/trainings?teamId=${teamId}&trainingDate=${trainingDate}&durationMinutes=${durationMinutes}&description=${description}&objective=${objective}`, config('POST'), onSuccess, onErrors);

export const addTrainingWithSeason = (seasonId, trainingDate, durationMinutes, description, objective, onSuccess, onErrors) =>
    appFetch(`/trainings?seasonId=${seasonId}&trainingDate=${trainingDate}&durationMinutes=${durationMinutes}&description=${description}&objective=${objective}`, config('POST'), onSuccess, onErrors);

export const addPlayerToTraining = (trainingId, playerId, onSuccess, onErrors) =>
    appFetch(`/trainings/${trainingId}/addPlayerToTraining?playerId=${playerId}`, config('POST'), onSuccess, onErrors);

export const updateTraining = (trainingId, trainingDate, durationMinutes, description, objective, onSuccess, onErrors) =>
    appFetch(`/trainings/${trainingId}?trainingDate=${trainingDate}&durationMinutes=${durationMinutes}&description=${description}&objective=${objective}`, config('PUT'), onSuccess, onErrors);

export const removeTraining = (trainingId, onSuccess, onErrors) =>
    appFetch(`/trainings/${trainingId}`, config('DELETE'), onSuccess, onErrors);

export const removePlayerToTraining = (playerId, trainingId, onSuccess, onErrors) =>
    appFetch(`/trainings/${playerId}/player?trainingId=${trainingId}`, config('DELETE'), onSuccess, onErrors);